File: TEAM A (Season 1, 2)/Week 3/Eren/150366. 표 병합/표 병합.cpp
#include <bits/stdc++.h>
#include <string.h>
using namespace std;
#define ii pair<int, int>

string board[54][54];

ii par[55][55];

ii Find(ii a){
    if(par[a.first][a.second] == a) return a;
    else return par[a.first][a.second] = Find(par[a.first][a.second]);
}


bool Union(ii a, ii b){
    a = Find({a.first, a.second});
    b = Find({b.first, b.second});
    
    if(a == b) return false;
    
    par[b.first][b.second] = a;
    return true;
}

vector<string> split(string input, string delimeter) {
    vector<string> ret;
    long long pos = 0;
    string token = "";
    while((pos = input.find(delimeter)) != string::npos) {
        token = input.substr(0, pos);
        ret.push_back(token);
        input.erase(0, pos + delimeter.length());
    }
    ret.push_back(input);
    return ret;
}

void reset(int x, int y){
    vector<ii> v;
    for(int i=1; i<=50; i++){
        for(int j=1; j<=50; j++){
            auto [a, b] = Find({i,j});
            if(x == a and y == b)
                v.push_back({i, j});
        }
    }
    for(auto [i, j]: v){
        par[i][j] = {i, j};
        board[i][j] = "";
    }
}


vector<string> solution(vector<string> commands) {
    vector<string> answer;
    
    for(int i=1; i<=50; i++)
        for(int j=1; j<=50; j++)
            par[i][j] = {i,j};
    
    for(auto command : commands) {
        vector<string> c = split(command, " ");
        int r1, c1, r2, c2;
        if(c[0] == "UPDATE" and c.size() == 4){
            r1 = stoi(c[1]); c1 = stoi(c[2]);
            
            auto [r3, c3] = Find({r1,c1});
            
            board[r3][c3] = c[3];
        }
        
        else if(c[0] == "UPDATE" and c.size() == 3){
            for(int i=1; i<=50; i++)
                for(int j=1; j<=50; j++)
                    if(board[i][j] == c[1])
                        board[i][j] = c[2];
        }
        
        else if(c[0] == "MERGE") {
            r1 = stoi(c[1]); c1 = stoi(c[2]);
            r2 = stoi(c[3]); c2 = stoi(c[4]);
            
            auto [r3, c3] = Find({r1, c1});
            auto [r4, c4] = Find({r2, c2});
            
            if(r3 == r4 and c3 == c4) continue;
            
            if(board[r3][c3] == "" && board[r4][c4] != ""){
                par[r3][c3] = {r4, c4};
            }
            else {
                par[r4][c4] = {r3, c3};
                board[r4][c4] = "";
            }
        }
        else if(c[0] == "UNMERGE"){
            r1 = stoi(c[1]); c1 = stoi(c[2]);
            
            auto [r3, c3] = Find({r1, c1});
            string tmp = board[r3][c3];
            
            reset(r3, c3);
            
            board[r1][c1] = tmp;
            
        }
        else if(c[0] == "PRINT"){
            r1 = stoi(c[1]); c1 = stoi(c[2]);
            
            auto [r3, c3] = Find({r1, c1});
            
            if(board[r3][c3] == "") answer.push_back("EMPTY");
            else answer.push_back(board[r3][c3]);
        }
        
    }
            
    return answer;
}

File: TEAM A (Season 1, 2)/Week 3/Jeckmu/150366. 표 병합/표 병합.py
table = [["EMPTY"]*51 for _ in range(51)]
ufTable = [[(r, c) for c in range(51)] for r in range(51)]
answer = []


def update(r, c, value):
    # 연결된 제일 상위 root 셀에 value 삽입.
    nr, nc = find(r, c)
    table[nr][nc] = value


def updateValue(v1, v2):
    # 모든 셀의 값을 검색하며 update.
    for i in range(51):
        for j in range(51):
            if table[i][j] == v1:
                table[i][j] = v2


def merge(r1, c1, r2, c2):
    nr1, nc1 = find(r1, c1)
    nr2, nc2 = find(r2, c2)

    if table[nr1][nc1] != "EMPTY":
        union(nr1, nc1, nr2, nc2)
    else:
        # 항상 (r1,c1)쪽으로 연결함과 동시에, (r2,c2)에만 값이 있다면 그 값을 (r1,c1)으로 옮겨오며 연결.
        if table[nr2][nc2] != "EMPTY":
            table[nr1][nc1] = table[nr2][nc2]

        union(nr1, nc1, nr2, nc2)


def unMerge(r, c):
    nr, nc = find(r, c)
    # 우선 param으로 받은 (r, c)셀에 연결된 root 셀에 값이 있다면, 저장해둠
    temp = table[nr][nc]

    tempList = []
    # 그 후, 연결된 셀들의 병합을 모두 해제 -> 모든 셀들의 root셀을 find하며, root셀이 같다면 연결 해제 후, 해제한 셀 값들은 초기화
    # 찾을 때마다 병합 해제를 하면, 연결된 모든 셀을 찾을 수 없으므로, 저장해뒀다가 한번에 해제
    for i in range(51):
        for j in range(51):
            if find(i, j) == (nr, nc):
                tempList.append((i, j))

    for t in tempList:
        i, j = t
        table[i][j] = "EMPTY"
        ufTable[i][j] = (i, j)

    # 아까 저장한 그 값을 (r, c)로 옮겨옴
    table[r][c] = temp


def print_(r, c):
    nr, nc = find(r, c)
    answer.append(table[nr][nc])


# Union-Find
def find(r, c):
    if (r, c) == ufTable[r][c]:
        return ufTable[r][c]

    nr, nc = ufTable[r][c]
    ufTable[r][c] = find(nr, nc)
    return ufTable[r][c]


def union(r1, c1, r2, c2):
    ufTable[r2][c2] = ufTable[r1][c1]


def solution(commands):
    for command in commands:
        c = list(command.split())

        if c[0] == "UPDATE":
            if len(c) == 4:
                # update r c value
                update(int(c[1]), int(c[2]), c[3])
            else:
                # update value1 value2
                updateValue(c[1], c[2])

        elif c[0] == "MERGE":
            merge(int(c[1]), int(c[2]), int(c[3]), int(c[4]))

        elif c[0] == "UNMERGE":
            unMerge(int(c[1]), int(c[2]))

        elif c[0] == "PRINT":
            print_(int(c[1]), int(c[2]))
    return answer